sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -22,137 +22,363 @@
+        for (int i = 0; i < matchers.length; i++) {
+            matchers[i] = ArgMatcher.rewriteTo(dataTypes.get(i));
+        }
+        return new VarArgs(matchers);
     }
 
-    public Signature(int varArgsStartingPosition, Collection<? extends DataType> c) {
-        assert varArgsStartingPosition < c.size() : "varArgs starting position exceeds args definition";
-        this.varArgsStartingPosition = varArgsStartingPosition;
-        list = ImmutableList.copyOf(c);
+    /**
+     * Convinience method for calling {@link #of(List<DataType>)} with varArgs.
+     */
+    public static SignatureOperator of(DataType... dataTypes) {
+        return of(Arrays.asList(dataTypes));
     }
 
-    public Signature(DataType... elements) {
-        this(-1, elements);
+    public interface SignatureOperator extends UnaryOperator<List<DataType>> {
+
+        /**
+         * Adds a precondition before the operator which should evaluate to true in order for the operator
+         * to be evaluated.
+         *
+         * @param predicate a predicate which represents the precondition
+         * @return a new operator with the predicate prepended
+         */
+        default SignatureOperator preCondition(Predicate<List<DataType>> predicate) {
+            return dataTypes -> predicate.test(dataTypes) ? this.apply(dataTypes) : null;
+        }
+
+        /**
+         * Adds a secondary operator to this operator, which is executed with the result of this operator if a match
+         * occurs.
+         *
+         * @param secondary the operator which will be evaluated with this operator's success result
+         * @return a new operator with the secondary appended
+         */
+        default SignatureOperator and(SignatureOperator secondary) {
+            return dataTypes -> {
+                List<DataType> n = this.apply(dataTypes);
+                if (n != null) {
+                    return secondary.apply(n);
+                }
+                return null;
+            };
+        }
+
     }
 
-    public Signature(int varArgsStartingPosition, DataType... elements) {
-        assert varArgsStartingPosition < elements.length : "varArgs starting position exceeds args definition";
-        this.varArgsStartingPosition = varArgsStartingPosition;
-        list = ImmutableList.copyOf(elements);
+    /**
+     * Interface for validating and normalizing an argument type of a function signature.
+     * <p>
+     * This interface is an unary operator for {@DataType} which returns a normalized type if the given input type
+     * is valid. Implementations can either pass through the input or replace the type if normalization is required.
+     * For example if the input type is null implementations might return a specific type if possible.
+     * <p>
+     * The operator returns null if the input does not match its criteria.
+     */
+    public interface ArgMatcher extends UnaryOperator<DataType> {
```

Comment:
ftr: decided to keep it because we couldn't come up with a better name than `apply`